packages/metrics-discovery/src/metrics-discovery.ts

```diff
@@ -0,0 +1,53 @@
+import type { NodeStructOutput } from '@river-build/generated/dev/typings/INodeRegistry'
+import { RiverRegistry, getWeb3Deployment } from '@river-build/web3'
+import { ethers } from 'ethers'
+
+const NODE_STATUS = {
+    ACTIVE: 2,
+    DEPARTING: 4,
+}
+
+// This class is in charge of pulling all nodes from river and
+// formatting them into a prometheus-friendly format
+
+export class MetricsDiscovery {
+    constructor(private readonly riverRegistry: RiverRegistry, private readonly env: string) {}
+
+    public static init(config: { riverRpcURL: string; env: string }) {
+        const deployment = getWeb3Deployment(config.env)
+        const provider = new ethers.providers.JsonRpcProvider(config.riverRpcURL)
+        const riverRegistry = new RiverRegistry(deployment.river, provider)
+        return new MetricsDiscovery(riverRegistry, config.env)
+    }
+
+    public static isTargeted(node: NodeStructOutput) {
+        return node.status === NODE_STATUS.ACTIVE || node.status === NODE_STATUS.DEPARTING
+    }
+
+    private async getTargetNodes() {
+        console.info('Getting target nodes')
+        const allNodes = await this.riverRegistry.nodeRegistry.read.getAllNodes()
```

Comment:
Just a future thought - As the number of nodes increases it may be cheaper to call `getAllNodes(nodeStatus)` if we are filtering on a subset of nodes, which we are.

Reply:
are you talking about riverRegistry.getAllNodes(status)? that one is basically the same idea as here: fetches all nodes from the contract and filters them at the client side. filtering by status doesn't exist on the contract level. will make a note to keep this in mind as the number of nodes grow though 👍